aro translate-c: support for record type translation

f-cozzocrea opened the pull request:
Edit: Ready for review. See below for a summary of changes.

---

This PR adds support for translating C record types. The PR is still very much a work in progress, so it's probably not worth reviewing now. The only tests that have been added are `simple_struct.c` and `simple_union.c`. 

Tasks to complete before it's ready:

- [x] Add more tests.
- [x] Cleanup/refactor.

ehaas commented on lib/compiler/aro_translate_c.zig:
```diff
@@ -668,6 +898,26 @@ fn transCreateNodeAPInt(c: *Context, int: aro.Value) !ZigNode {
     return res;
 }
 
+fn cIsStruct(ty: Type) bool {
```
No need for these - you can use `ty.is(.@"struct")` / `ty.is(.@"union")`

f-cozzocrea replied:
Thanks for the heads up, that's good to know. Updated.

f-cozzocrea commented on lib/compiler/aro/aro/Type.zig:
```diff
@@ -1142,12 +1142,14 @@ pub fn alignof(ty: Type, comp: *const Compilation) u29 {
 }
 
+pub const QualHandling = enum { standard, preserve_quals };
+
 /// Canonicalize a possibly-typeof() type. If the type is not a typeof() type, simply
 /// return it. Otherwise, determine the actual qualified type.
 /// The `qual_handling` parameter can be used to return the full set of qualifiers
 /// added by typeof() operations, which is useful when determining the elemType of
 /// arrays and pointers.
-pub fn canonicalize(ty: Type, qual_handling: enum { standard, preserve_quals }) Type {
+pub fn canonicalize(ty: Type, qual_handling: QualHandling) Type {
```
This is one change that I made to the aro source which makes qual handling in type translation a bit easier. If this change is acceptable, I'll submit an upstream PR with this change.

f-cozzocrea commented:
This PR ended up being bigger than I thought it'd be, so there are a few features unimplemented. This is my first time working with the aro codebase, so feedback and pointers are welcome. Here's a summary of the changes:

1. `fn locStr()` fixed to report the correct C source location. In order to facilitate this, I added a `source: aro.Source` field to `Context`. It looks like `Driver` is the correct place to get the `Source` value rather than `Comp`, but let me know if this is incorrect.
2. Moved tests related to record types from the old test harness to the new, and added a few new tests. Unfortunately, I wasn't able to add all of the tests because many of them have features (eg. pointers) that aren't supported by aro translate-c.
3. As indicated in my above comment, I made a small change to make the qualifier handling enum for `fn canonicalize()` a public type. This was done for passing the enum value to the `transType()` function.
4. Changed the behaviour of `prepopulateGlobalNameTable()` to add enums and records to the `weak_global_names` table, and all other names to the `global_names` table. This matches the behaviour of Clang-based translate-c.

ehaas commented on lib/compiler/aro_translate_c.zig:
```diff
@@ -215,80 +221,251 @@ fn prepopulateGlobalNameTable(c: *Context) !void {
+    const node_types = c.tree.nodes.items(.ty);
+    const raw_record_ty = node_types[@intFromEnum(record_node)];
+    const ty = raw_record_ty.canonicalize(.preserve_quals);
+    const record_decl = ty.data.record;
```
```suggestion
    const record_decl = raw_record_ty.getRecord().?;
```

f-cozzocrea replied:
Added.

ehaas commented on lib/compiler/aro_translate_c.zig:
```diff
@@ -215,80 +221,251 @@ fn prepopulateGlobalNameTable(c: *Context) !void {
+fn transRecordDecl(c: *Context, scope: *Scope, record_node: NodeIndex, field_nodes: []const NodeIndex) Error!void {
+    const node_types = c.tree.nodes.items(.ty);
+    const raw_record_ty = node_types[@intFromEnum(record_node)];
+    const ty = raw_record_ty.canonicalize(.preserve_quals);
```
I don't think `ty` is needed, you can just use `raw_record_ty` with the `is` predicates

f-cozzocrea replied:
I like that. `ty` removed.

ehaas commented on lib/compiler/aro_translate_c.zig:
```diff
@@ -215,80 +221,251 @@ fn prepopulateGlobalNameTable(c: *Context) !void {
+
+        var unnamed_field_count: u32 = 0;
+
+        for (record_decl.fields, field_nodes, 0..) |field, field_node, i| {
```
`record_decl.fields` and `field_nodes` won't necessarily have the same length; the former is the fields; the latter in this case is all the nodes inside that struct, which might include e.g. anonymous record decls. I think you'll need 2 loops, one over the nodes and one over the fields. Any nodes that are decls should be translated. For the fields loop, you can use `field.ty` for the type instead of looking it up in the node types slice. I also don't think you have to canonicalize the field type.

f-cozzocrea replied:
Ah, that makes sense. Thanks for explaining that, I'll update that.

ehaas commented on lib/compiler/aro_translate_c.zig:
```diff
@@ -215,80 +221,251 @@ fn prepopulateGlobalNameTable(c: *Context) !void {
+            if (field_qt.isEnumOrRecord()) try transDecl(c, scope, field_node);
+            const field_type = transType(c, scope, field_qt, .preserve_quals, field_loc) catch |err| switch (err) {
+                error.UnsupportedType => {
+                    try c.opaque_demotes.put(c.gpa, @intFromPtr(ty.canonicalize(.standard).data.record), {});
```
```suggestion
                    try c.opaque_demotes.put(c.gpa, @intFromPtr(record_decl), {});
```

ehaas commented on lib/compiler/aro_translate_c.zig:
```diff
@@ -496,16 +673,59 @@ fn transType(c: *Context, scope: *Scope, raw_ty: Type, source_loc: TokenIndex) T
+// u29 after the Clang frontend for translate-c is removed. Note that the raw type should be
+// passed to detect attributes that would affect alignment.
+fn alignmentForField(c: *Context, field_ty: Type, parent: Type) ?c_uint {
+    const parent_packed = if (parent.specifier == .attributed) parent.hasAttribute(.@"packed") else false;
```
```suggestion
    const parent_packed = parent.hasAttribute(.@"packed");
```

ehaas commented on lib/compiler/aro_translate_c.zig:
```diff
@@ -215,80 +221,251 @@ fn prepopulateGlobalNameTable(c: *Context) !void {
+                else => |e| return e,
+            };
+
+            var field_ty_with_attrs = field.ty;
```
We compute record layouts so I don't think this is needed; `field.layout` tells us the field's offset in bits from the start of the record. Also note that the requested alignment is not always granted by the compiler, but the record layout code accounts for that.

f-cozzocrea replied:
I see, that makes sense. Getting the requested attribute isn't the way to go then. I'm not sure we can get the alignment from just the field offset though. For example, how would this test case work? The offset is 0 with or without the attribute. Would we use one of the values in `TypeLayout`?

```c
// struct_with_aligned_field.c
struct foo {
    __attribute__((aligned(1))) short bar;
};
```

```zig
// translated zig code
pub const struct_foo = extern struct {
    bar: c_short align(1) = @import("std").mem.zeroes(c_short),
};
```

ehaas replied:
Sorry, should have mentioned the record's `type_layout` as well. The example you posted seems like a bug in current translate-c - in C the `aligned` attribute can only increase a field's alignment, not decrease it (unless the record is packed).

```c
// x86_64-linux
struct foo {
    __attribute__((aligned(1))) short bar;
};
_Static_assert(_Alignof(struct foo) == 2, "");
```
whereas the translated zig struct you pasted will have an alignment of 1.

I believe you want the `pointer_alignment_bits` of the record's `type_layout`, which tells you the required alignment of valid pointers to that type.

f-cozzocrea replied:
Okay, that makes sense! I'll play around with those fields and see what I can come up with. 

I forgot that the aligned attribute can only increase the alignment. I think that's a bug too. The clang translate-c is only looking at the attributes, not the actual record layout. I'll extract this and make a tracking issue.

Vexu reviewed (commented):
Nice work so far! I'll leave the layout stuff for you and ehaas for now

f-cozzocrea commented on lib/compiler/aro_translate_c.zig:
```diff
@@ -215,80 +221,254 @@ fn prepopulateGlobalNameTable(c: *Context) !void {
+        bare_name = typedef_name;
+        name = typedef_name;
+    } else {
+        // Record declarations such as `struct {...} x` have no name but they're not
+        // anonymous hence here isAnonymousStructOrUnion is not needed
```
Thanks, forgot to strip that when referencing the clang translate-c.

f-cozzocrea commented on lib/compiler/aro_translate_c.zig:
```diff
@@ -215,80 +221,254 @@ fn prepopulateGlobalNameTable(c: *Context) !void {
+
+        var unnamed_field_count: u32 = 0;
+
+        // Iterate over field nodes so that translate any type decls included in the record decl.
```
I'm not quite sure what you mean, but it's not a temporary mechanism. This is meant to be similar to the behaviour of the current translate-c code. The difference is that the clang-based translate-c does the decl translation inside of `transType()`.

I can certainly make the comment more clear about the intention of the for loop. Basically, the for loop covers cases where an enum or record is declared inside of the record declaration. Eg:
```C
struct foo {
    union {
        int bar;
        long baz;
    };
};
```
 Does that answer your question?

Vexu replied:
> The difference is that the clang-based translate-c does the decl translation inside of `transType()`.

The Aro based one should be doing that as well. But this is a perfectly valid temporary solution as long as it is documented as such.

ehaas commented:
@f-cozzocrea I notice your last commit says "needs cleanup" - do you want more feedback now, or after the cleanup?

f-cozzocrea commented:
@ehaas After the cleanup. After sleeping on it, there's a few things I want to change about the approach before it's ready.

f-cozzocrea commented:
@ehaas I pushed some changes, and I feel like the code is in a good place to get feedback. One aspect I'd like feedback about is the alignment of zero-width fields. I have it set to translate to an alignment of 1 right now because that's the behaviour of `alignof()` in C, but I'm not sure if that's the correct value for translation.

ehaas commented on lib/compiler/aro_translate_c.zig:
```diff
@@ -551,7 +874,7 @@ fn transFnType(
     };
 
-    const alignment = if (raw_ty.requestedAlignment(c.comp)) |alignment| zigAlignment(alignment) else null;
+    const alignment = alignmentForFn(c, raw_ty);
```
```suggestion
    const alignment: ?c_uint = raw_ty.requestedAlignment(c.comp) orelse null;
```

ehaas commented on lib/compiler/aro_translate_c.zig:
```diff
@@ -496,16 +667,168 @@ fn transType(c: *Context, scope: *Scope, raw_ty: Type, source_loc: TokenIndex) T
+        // Records have a natural alignment when used as a field, and their size is
+        // a multiple of this alignment value. For all other types, the natural alignment
+        // is their size.
+        const field_natural_alignment_bits: u64 = blk: {
```
```suggestion
        const field_natural_alignment_bits = if (field.ty.getRecord()) |record| record.type_layout.field_alignment_bits else field_size_bits;
```

ehaas reviewed (changes requested):
It looks like the `alignmentForFields` function could calculate the alignment for an individual field given a small fixed amount of context (the `max_field_alignment_bits` for the record + the calculated alignment of the previous field, if any). That way you could compute the alignments on the fly for each field, and not have to allocate a slice of `?c_uint`

You can also avoid even having to manually calculate alignments if the record has no attributes - something like:
```zig
 const is_simple = (raw_record_ty.specifier == .@"union" or raw_record_ty.specifier == .@"struct") and record_decl.field_attributes == null;
```
If there are no attributes, the default layout provided by extern struct/union should be sufficient.

ehaas commented:
I think an alignment of 1 for zero-width fields is fine - those fields aren't really usable, they're just to make subsequent fields align differently, so as long as those following fields are aligned I don't see any problems. Perhaps there are more subtle issues at play - @Vexu does that seem ok to you?

f-cozzocrea commented:
Thanks for the feedback @ehaas, both of those comments make sense to me.

ehaas commented on lib/compiler/aro_translate_c.zig:
```diff
@@ -496,16 +711,126 @@ fn transType(c: *Context, scope: *Scope, raw_ty: Type, source_loc: TokenIndex) T
 
-fn zigAlignment(bit_alignment: u29) u32 {
-    return bit_alignment / 8;
+// This function returns an array of ?c_uint to match Clang's behaviour of using c_uint.
```
This comment is out of date. Also let's make it a doc comment (`///` instead of `//`)

ehaas commented on lib/compiler/aro_translate_c.zig:
```diff
@@ -496,16 +711,125 @@ fn transType(c: *Context, scope: *Scope, raw_ty: Type, source_loc: TokenIndex) T
+    head_field_alignment: ?c_uint,
+    field_index: usize,
+) ?c_uint {
+    const record_decl = record_ty.getRecord().?;
```
Sorry - one other thing I just noticed - you can just pass the record decl in directly instead of having to call `getRecord` each time in this function - it looks like `record_ty` isn't used for anything else.

f-cozzocrea replied:
Oh that's a good catch! Updating it now.

ehaas commented on lib/compiler/aro_translate_c.zig:
```diff
@@ -215,80 +218,292 @@ fn prepopulateGlobalNameTable(c: *Context) !void {
+            else
+                null;
+
+            try fields.append(.{
```
`fields` was created with `initCapacity` of the number of fields, so you should be able to do `appendAssumeCapacity` here.

ehaas commented on lib/compiler/aro_translate_c.zig:
```diff
@@ -215,80 +218,292 @@ fn prepopulateGlobalNameTable(c: *Context) !void {
+        const has_alignment_attributes = record_decl.field_attributes != null or
+            raw_record_ty.hasAttribute(.@"packed") or
+            raw_record_ty.hasAttribute(.aligned);
+        const head_field_alignment: ?c_uint = head_align: {
```
Bikeshed/personal preference but this logic might be better off extracted into its own function to make the `transRecordDecl` function more readable.

ehaas replied:
Specifically referring to the `head_field_alignment` calculation

f-cozzocrea replied:
I think that's a good idea. I'm usually reluctant to add functions that are only called once, but this definitely helps readability.

ehaas commented on lib/compiler/aro_translate_c.zig:
```diff
@@ -215,80 +218,292 @@ fn prepopulateGlobalNameTable(c: *Context) !void {
+    const toplevel = scope.id == .root;
+    const bs: *Scope.Block = if (!toplevel) try scope.findBlockScope(c) else undefined;
+
+    var is_union = false;
```
```suggestion
    const container_kind: ZigTag = if (raw_record_ty.is(.@"union")) .@"union" else .@"struct";
```

Then `container_kind_name` usage can be replaced with `@tagName(container_kind)`; the record payload `{ .tag = ([2]ZigTag{ .@"struct", .@"union" })[@intFromBool(is_union)] }` can be replaced with `.{ .tag = container_kind }`, and the `default_value` node condition can check `container_kind != .@"union"`

ehaas reviewed (commented):
A few more comments and then I think this will be in really good shape! Nice work!

f-cozzocrea commented:
Rebased on master and squashed to 1 commit.

(Sorry for the review request Snektron. Accidentally borked the rebase the first time.)

f-cozzocrea commented:
> A few more comments and then I think this will be in really good shape! Nice work!

Awesome! Thanks for the feedback to make this a high quality PR. 😄

ehaas commented on lib/compiler/aro_translate_c.zig:
```diff
@@ -215,80 +218,256 @@ fn prepopulateGlobalNameTable(c: *Context) !void {
+        const has_alignment_attributes = record_decl.field_attributes != null or
+            raw_record_ty.hasAttribute(.@"packed") or
+            raw_record_ty.hasAttribute(.aligned);
+        const head_field_alignment: ?c_uint = headFieldAlignment(record_decl);
```
`head_field_alignment` isn't used if `has_alignment_attributes` is false, so if that's the case we should just initialized it to null instead of computing a value via `headFieldAlignment`.

Vexu commented:
Absolutely, sorry I forgot about this, I'll get on it in tomorrow.

Vexu commented on lib/compiler/aro_translate_c.zig:
```diff
@@ -496,16 +675,152 @@ fn transType(c: *Context, scope: *Scope, raw_ty: Type, source_loc: TokenIndex) T
+    }
+}
+
+/// This function returns a ?c_uint to match Clang's behaviour of using c_uint.
+/// This can be changed to a u29 after the Clang frontend for translate-c is removed.
```
Zig alignment is going to be `u16` in the future so we might as well use that here.

f-cozzocrea replied:
Doesn't that mean that C code with alignments wider than `u16` will be untranslatable to Zig? If so, I'd also need to add code to demote variables and functions to `extern` if there's an alignment value that's too high.

Vexu commented on lib/compiler/aro_translate_c.zig:
```diff
@@ -496,16 +675,152 @@ fn transType(c: *Context, scope: *Scope, raw_ty: Type, source_loc: TokenIndex) T
+
+/// This function returns a ?c_uint to match Clang's behaviour of using c_uint.
+/// This can be changed to a u29 after the Clang frontend for translate-c is removed.
+fn alignmentForField(
```
It looks like the reason this function is complicated is because of differences in alignment in Zig and C but I'd like it to document that.

Vexu reviewed (approved):
Looks good aside from some small changes to one function! Sorry for the delay in getting this reviewed.

f-cozzocrea commented:
No worries. I figured there'd be a delay because of the 0.12 release. I'm in the middle of moving, so I'll look at your comments and address them in the next week or so.

andrewrk commented:
Looks like @Vexu considers this good enough to be merged as-is, so those follow-up improvements can be sent later. Nice work!
